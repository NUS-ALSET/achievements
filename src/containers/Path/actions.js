export const PATH_OPEN = "PATH_OPEN";
export const pathOpen = pathId => ({
  type: PATH_OPEN,
  pathId
});

export const PATH_OPEN_SOLUTION_DIALOG = "PATH_OPEN_SOLUTION_DIALOG";
export const pathOpenSolutionDialog = (pathId, problemInfo) => ({
  type: PATH_OPEN_SOLUTION_DIALOG,
  pathId,
  problemInfo
});

export const PATH_CLOSE_DIALOG = "PATH_CLOSE_DIALOG";
export const pathCloseDialog = () => ({
  type: PATH_CLOSE_DIALOG
});

// In fact, it only fetches existance of solution, should be renamed?
export const PATH_FETCH_PROBLEMS_SOLUTIONS_SUCCESS =
  "PATH_FETCH_PROBLEMS_SOLUTIONS_SUCCESS";
export const pathFetchProblemsSolutionsSuccess = (pathId, solutions) => ({
  type: PATH_FETCH_PROBLEMS_SOLUTIONS_SUCCESS,
  pathId,
  solutions
});

export const PATH_MORE_PROBLEMS_REQUEST = "PATH_MORE_PROBLEMS_REQUEST";
export const pathMoreProblemsRequest = (userId, pathId, activityCount) => ({
  type: PATH_MORE_PROBLEMS_REQUEST,
  userId,
  pathId,
  activityCount
});

export const PATH_SHOW_COLLABORATORS_DIALOG = "PATH_SHOW_COLLABORATORS_DIALOG";
export const pathShowCollaboratorsDialog = pathId => ({
  type: PATH_SHOW_COLLABORATORS_DIALOG,
  pathId
});

export const PATH_COLLABORATORS_FETCH_SUCCESS =
  "PATH_COLLABORATORS_FETCH_SUCCESS";
export const pathCollaboratorsFetchSuccess = (pathId, assistants) => ({
  type: PATH_COLLABORATORS_FETCH_SUCCESS,
  pathId,
  assistants
});

export const PATH_ADD_COLLABORATOR_REQUEST = "PATH_ADD_COLLABORATOR_REQUEST";
export const pathAddCollaboratorRequest = (pathId, collaboratorId) => ({
  type: PATH_ADD_COLLABORATOR_REQUEST,
  pathId,
  collaboratorId
});

export const PATH_ADD_COLLABORATOR_SUCCESS = "PATH_ADD_COLLABORATOR_SUCCESS";
export const pathAddCollaboratorSuccess = (pathId, collaboratorId) => ({
  type: PATH_ADD_COLLABORATOR_SUCCESS,
  pathId,
  collaboratorId
});

export const PATH_ADD_COLLABORATOR_FAIL = "PATH_ADD_COLLABORATOR_FAIL";
export const pathAddCollaboratorFail = (pathId, collaboratorId, reason) => ({
  type: PATH_ADD_COLLABORATOR_FAIL,
  pathId,
  collaboratorId,
  reason
});

export const PATH_REMOVE_COLLABORATOR_REQUEST =
  "PATH_REMOVE_COLLABORATOR_REQUEST";
export const pathRemoveCollaboratorRequest = (pathId, collaboratorId) => ({
  type: PATH_REMOVE_COLLABORATOR_REQUEST,
  pathId,
  collaboratorId
});

export const PATH_REMOVE_COLLABORATOR_SUCCESS =
  "PATH_REMOVE_COLLABORATOR_SUCCESS";
export const pathRemoveCollaboratorSuccess = (pathId, collaboratorId) => ({
  type: PATH_REMOVE_COLLABORATOR_SUCCESS,
  pathId,
  collaboratorId
});

export const PATH_REMOVE_COLLABORATOR_FAIL = "PATH_REMOVE_COLLABORATOR_FAIL";
export const pathRemoveCollaboratorFail = (pathId, collaboratorId, reason) => ({
  type: PATH_REMOVE_COLLABORATOR_FAIL,
  pathId,
  collaboratorId,
  reason
});

export const PATH_MORE_PROBLEMS_SUCCESS = "PATH_MORE_PROBLEMS_SUCCESS";
export const pathMoreProblemsSuccess = (userId, pathId, activityCount) => ({
  type: PATH_MORE_PROBLEMS_SUCCESS,
  userId,
  pathId,
  activityCount
});

export const PATH_MORE_PROBLEMS_FAIL = "PATH_MORE_PROBLEMS_FAIL";
export const pathMoreProblemsFail = (
  userId,
  pathId,
  activityCount,
  reason
) => ({
  type: PATH_MORE_PROBLEMS_FAIL,
  userId,
  pathId,
  activityCount,
  reason
});

export const PATH_TOGGLE_JOIN_STATUS_REQUEST =
  "PATH_TOGGLE_JOIN_STATUS_REQUEST";
export const pathToggleJoinStatusRequest = (userId, pathId, status) => ({
  type: PATH_TOGGLE_JOIN_STATUS_REQUEST,
  userId,
  pathId,
  status
});

export const PATH_TOGGLE_JOIN_STATUS_SUCCESS =
  "PATH_TOGGLE_JOIN_STATUS_SUCCESS";
export const pathToggleJoinStatusSuccess = (pathId, status) => ({
  type: PATH_TOGGLE_JOIN_STATUS_SUCCESS,
  pathId,
  status
});

export const PATH_TOGGLE_JOIN_STATUS_FAIL = "PATH_TOGGLE_JOIN_STATUS_FAIL";
export const pathToggleJoinStatusFail = (pathId, status, reason) => ({
  type: PATH_TOGGLE_JOIN_STATUS_FAIL,
  pathId,
  status,
  reason
});

export const PATH_PROBLEM_OPEN = "PATH_PROBLEM_OPEN";
export const pathProblemOpen = (pathId, problemId) => ({
  type: PATH_PROBLEM_OPEN,
  pathId,
  problemId
});

export const PATH_REFRESH_SOLUTIONS_REQUEST = "PATH_REFRESH_SOLUTIONS_REQUEST";
export const pathRefreshSolutionsRequest = pathId => ({
  type: PATH_REFRESH_SOLUTIONS_REQUEST,
  pathId
});

export const PATH_REFRESH_SOLUTIONS_SUCCESS = "PATH_REFRESH_SOLUTIONS_SUCCESS";
export const pathRefreshSolutionsSuccess = pathId => ({
  type: PATH_REFRESH_SOLUTIONS_SUCCESS,
  pathId
});

export const PATH_REFRESH_SOLUTIONS_FAIL = "PATH_REFRESH_SOLUTIONS_FAIL";
export const pathRefreshSolutionsFail = (pathId, reason) => ({
  type: PATH_REFRESH_SOLUTIONS_FAIL,
  pathId,
  reason
});
