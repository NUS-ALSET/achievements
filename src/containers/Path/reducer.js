import {
  PATH_ADD_COLLABORATOR_SUCCESS,
  PATH_CLOSE_DIALOG,
  PATH_COLLABORATORS_FETCH_SUCCESS,
  PATH_FETCH_PROBLEMS_SOLUTIONS_SUCCESS,
  PATH_OPEN_SOLUTION_DIALOG,
  PATH_REMOVE_COLLABORATOR_SUCCESS,
  PATH_SHOW_COLLABORATORS_DIALOG
} from "./actions";
import { PATH_PROBLEM_DIALOG_SHOW } from "../Paths/actions";
import { ASSIGNMENT_ASSISTANT_FOUND } from "../Assignments/actions";

export const path = (
  state = {
    problemSolutions: {},
    ui: {
      dialog: {
        type: ""
      }
    }
  },
  action
) => {
  switch (action.type) {
    case PATH_CLOSE_DIALOG:
      return {
        ...state,
        ui: {
          dialog: {
            type: ""
          }
        }
      };
    case PATH_OPEN_SOLUTION_DIALOG:
      return {
        ...state,
        ui: {
          dialog: {
            type: action.problemInfo.type + "Solution",
            value: action.problemInfo
          }
        }
      };
    case PATH_PROBLEM_DIALOG_SHOW:
      return {
        ...state,
        ui: {
          ...state.ui,
          dialog: {
            type: "ProblemChange",
            value: action.problemInfo
          }
        }
      };
    case PATH_FETCH_PROBLEMS_SOLUTIONS_SUCCESS:
      return {
        ...state,
        problemSolutions: Object.assign(
          {},
          state.problemSolutions,
          action.solutions
        )
      };
    case PATH_COLLABORATORS_FETCH_SUCCESS:
      return {
        ...state,
        ui: {
          ...state.ui,
          dialog: {
            ...state.ui.dialog,
            assistants: action.assistants
          }
        }
      };
    case ASSIGNMENT_ASSISTANT_FOUND:
      return {
        ...state,
        ui: {
          ...state.ui,
          dialog: {
            ...state.ui.dialog,
            newAssistant: action.assistant
          }
        }
      };
    case PATH_ADD_COLLABORATOR_SUCCESS:
      return {
        ...state,
        ui: {
          dialog: {
            ...state.ui.dialog,
            assistants: [
              ...(state.ui.dialog.assistants || []),
              state.ui.dialog.newAssistant
            ]
          }
        }
      };
    case PATH_REMOVE_COLLABORATOR_SUCCESS:
      return {
        ...state,
        ui: {
          dialog: {
            ...state.ui.dialog,
            assistants: state.ui.dialog.assistants.filter(
              assistant => assistant.id !== action.collaboratorId
            )
          }
        }
      };
    case PATH_SHOW_COLLABORATORS_DIALOG:
      return {
        ...state,
        ui: {
          ...state.ui,
          dialog: {
            type: "CollaboratorsControl",
            assistants: false
          }
        }
      };
    default:
      return state;
  }
};
