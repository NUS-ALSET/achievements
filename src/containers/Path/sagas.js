import { call, put, select, take, takeLatest } from "redux-saga/effects";
import {
  PATH_ADD_COLLABORATOR_REQUEST,
  PATH_MORE_PROBLEMS_REQUEST,
  PATH_OPEN,
  PATH_PROBLEM_OPEN,
  PATH_REFRESH_SOLUTIONS_REQUEST,
  PATH_REMOVE_COLLABORATOR_REQUEST,
  PATH_SHOW_COLLABORATORS_DIALOG,
  PATH_TOGGLE_JOIN_STATUS_REQUEST,
  pathAddCollaboratorFail,
  pathAddCollaboratorSuccess,
  pathCollaboratorsFetchSuccess,
  pathMoreProblemsFail,
  pathMoreProblemsSuccess,
  pathRefreshSolutionsFail,
  pathRefreshSolutionsSuccess,
  pathRemoveCollaboratorFail,
  pathRemoveCollaboratorSuccess,
  pathToggleJoinStatusFail,
  pathToggleJoinStatusRequest,
  pathToggleJoinStatusSuccess
} from "./actions";
import { pathsService } from "../../services/paths";
import {
  PATH_ACTIVITY_MOVE_REQUEST,
  pathActivityMoveFail,
  PATHS_JOINED_FETCH_SUCCESS
} from "../Paths/actions";
import { notificationShow } from "../Root/actions";
import { codeCombatProfileSelector, pathActivitiesSelector } from "./selectors";

export function* pathProblemOpenHandler(action) {
  const data = yield select(state => ({
    uid: state.firebase.auth.uid,
    joinedPaths: state.paths.joinedPaths
  }));

  if (!data.joinedPaths[action.pathId]) {
    yield put(pathToggleJoinStatusRequest(data.uid, action.pathId, true));
  }
}

export function* pathOpenHandler(action) {
  if (!action.pathId || action.pathId[0] !== "-") {
    return yield Promise.resolve();
  }

  let owner = yield select(
    state =>
      state.paths &&
      state.paths.joinedPaths &&
      state.paths.joinedPaths[action.pathId] &&
      state.paths.joinedPaths[action.pathId].owner
  );
  if (!owner) {
    yield take(PATHS_JOINED_FETCH_SUCCESS);
    yield select(
      state =>
        state.paths &&
        state.paths.joinedPaths &&
        state.paths.joinedPaths[action.pathId] &&
        state.paths.joinedPaths[action.pathId].owner
    );
  }
}

export function* pathToggleJoinStatusRequestHandler(action) {
  try {
    const uid = yield select(state => state.firebase.auth.uid);
    const path = yield call(
      pathsService.togglePathJoinStatus,
      uid,
      action.pathId,
      action.status
    );
    yield put(
      pathToggleJoinStatusSuccess(action.pathId, action.status && path)
    );
  } catch (err) {
    yield put(
      pathToggleJoinStatusFail(action.pathId, action.status, err.message)
    );
  }
}

export function* pathActivityMoveRequestHandler(action) {
  try {
    const data = yield select(state => ({
      uid: state.firebase.auth.uid,
      activities: state.firebase.data.activities || {}
    }));

    const activities = Object.keys(data.activities)
      .map(id => ({ ...data.activities[id], id }))
      .filter(activity => activity.path === action.pathId);

    yield call(
      [pathsService, pathsService.moveActivity],
      data.uid,
      action.pathId,
      activities,
      action.activityId,
      action.direction
    );
  } catch (err) {
    yield put(
      pathActivityMoveFail(
        action.pathId,
        action.activityId,
        action.direction,
        err.message
      )
    );
    yield put(notificationShow(err.message));
  }
}

export function* pathMoreProblemsRequestHandler(action) {
  try {
    yield call(
      pathsService.storeMoreProblemsRequest,
      action.userId,
      action.pathId,
      action.activityCount
    );
    yield put(
      pathMoreProblemsSuccess(
        action.userId,
        action.pathId,
        action.activityCount
      )
    );
  } catch (err) {
    yield put(
      pathMoreProblemsFail(
        action.userId,
        action.pathId,
        action.activityCount,
        err.message
      )
    );
  }
}

/**
 * Note: there is names "collision" - `collaborators` at courses called as
 * `assistants`, so, somewhere I tried to keep consistence, somewhere I lost it
 * @param action
 * @returns {IterableIterator<*>}
 */
export function* pathShowCollaboratorsDialogHandler(action) {
  try {
    const collaborators = yield call(
      pathsService.fetchPathCollaborators,
      action.pathId
    );
    yield put(pathCollaboratorsFetchSuccess(action.pathId, collaborators));
  } catch (err) {
    yield put(notificationShow(err.message));
  }
}

export function* pathAddCollaboratorRequestHandler(action) {
  try {
    yield call(
      pathsService.updatePathCollaborator,
      action.pathId,
      action.collaboratorId,
      "add"
    );
    yield put(pathAddCollaboratorSuccess(action.pathId, action.collaboratorId));
  } catch (err) {
    yield put(
      pathAddCollaboratorFail(action.pathId, action.collaboratorId, err.message)
    );
    yield put(notificationShow(err.message));
  }
}

export function* pathRemoveCollaboratorRequestHandler(action) {
  try {
    yield call(
      pathsService.updatePathCollaborator,
      action.pathId,
      action.collaboratorId,
      "remove"
    );
    yield put(
      pathRemoveCollaboratorSuccess(action.pathId, action.collaboratorId)
    );
  } catch (err) {
    yield put(
      pathRemoveCollaboratorFail(
        action.pathId,
        action.collaboratorId,
        err.message
      )
    );
    yield put(notificationShow(err.message));
  }
}

export function* pathRefreshSolutionsRequestHandler(action) {
  try {
    const data = yield select(state => ({
      uid: state.firebase.auth.uid,
      pathActivities: pathActivitiesSelector(state, {
        match: { params: action }
      }),
      profile: codeCombatProfileSelector(state)
    }));

    yield call(
      [pathsService, pathsService.refreshPathSolutions],
      data.uid,
      data.pathActivities,
      data.profile
    );
    yield put(pathRefreshSolutionsSuccess(action.pathId));
  } catch (err) {
    yield put(pathRefreshSolutionsFail(action.pathId, err.message));
    yield put(notificationShow(err.message));
  }
}

export default [
  function* watchPathOpenRequest() {
    yield takeLatest(PATH_OPEN, pathOpenHandler);
  },
  function* watchPathProblemOpen() {
    yield takeLatest(PATH_PROBLEM_OPEN, pathProblemOpenHandler);
  },
  function* watchPathToggleJoinStatusRequest() {
    yield takeLatest(
      PATH_TOGGLE_JOIN_STATUS_REQUEST,
      pathToggleJoinStatusRequestHandler
    );
  },
  function* watchPathActivityMoveRequest() {
    yield takeLatest(
      PATH_ACTIVITY_MOVE_REQUEST,
      pathActivityMoveRequestHandler
    );
  },
  function* watchPathMoreProblemsRequest() {
    yield takeLatest(
      PATH_MORE_PROBLEMS_REQUEST,
      pathMoreProblemsRequestHandler
    );
  },

  function* watchPathShowCollaboratorsDialog() {
    yield takeLatest(
      PATH_SHOW_COLLABORATORS_DIALOG,
      pathShowCollaboratorsDialogHandler
    );
  },
  function* watchPathAddCollaboratorRequest() {
    yield takeLatest(
      PATH_ADD_COLLABORATOR_REQUEST,
      pathAddCollaboratorRequestHandler
    );
  },
  function* watchPathRemoveCollaboratorRequest() {
    yield takeLatest(
      PATH_REMOVE_COLLABORATOR_REQUEST,
      pathRemoveCollaboratorRequestHandler
    );
  },
  function* watchPathRefreshSolutionsRequest() {
    yield takeLatest(
      PATH_REFRESH_SOLUTIONS_REQUEST,
      pathRefreshSolutionsRequestHandler
    );
  }
];
