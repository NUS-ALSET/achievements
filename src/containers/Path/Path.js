/**
 * @file Path container module
 * @author Theodor Shaytanov <theodor.shaytanov@gmail.com>
 * @created 17.03.18
 */
import React, { Fragment } from "react";
import PropTypes from "prop-types";
import { compose } from "redux";
import { connect } from "react-redux";
import { push } from "connected-react-router";

import { APP_SETTING } from "../../achievementsApp/config";

import withStyles from "@material-ui/core/styles/withStyles";

import Button from "@material-ui/core/Button";
import Toolbar from "@material-ui/core/Toolbar";
import LinearProgress from "@material-ui/core/LinearProgress";

import ActivitiesTable from "../../components/tables/ActivitiesTable";

import { firebaseConnect } from "react-redux-firebase";

import withRouter from "react-router-dom/withRouter";
import Breadcrumbs from "../../components/Breadcrumbs";
import {
  pathStatusSelector,
  pathActivitiesSelector,
  PATH_STATUS_JOINED,
  PATH_STATUS_OWNER,
  PATH_STATUS_NOT_JOINED,
  codeCombatProfileSelector
} from "./selectors";
import {
  pathAddCollaboratorRequest,
  pathCloseDialog,
  pathMoreProblemsRequest,
  pathOpen,
  pathOpenSolutionDialog,
  pathRefreshSolutionsRequest,
  pathRemoveCollaboratorRequest,
  pathShowCollaboratorsDialog,
  pathToggleJoinStatusRequest
} from "./actions";
import {
  pathProblemChangeRequest,
  pathProblemDialogShow,
  pathActivityMoveRequest
} from "../Paths/actions";
import ActivityDialog from "../../components/dialogs/ActivityDialog";

import AddIcon from "@material-ui/icons/Add";
import { sagaInjector } from "../../services/saga";
import sagas from "./sagas";
import AddTextSolutionDialog from "../../components/dialogs/AddTextSolutionDialog";
import AddJestSolutionDialog from "../../components/dialogs/AddJestSolutionDialog";
import { ACTIVITY_TYPES } from "../../services/paths";
import { notificationShow } from "../Root/actions";
import { problemSolutionSubmitRequest } from "../Activity/actions";
import AddProfileDialog from "../../components/dialogs/AddProfileDialog";
import { externalProfileUpdateRequest } from "../Account/actions";
import { pathActivities } from "../../types";
import ControlAssistantsDialog from "../../components/dialogs/ControlAssistantsDialog";
import { assignmentAssistantKeyChange } from "../Assignments/actions";

const styles = theme => ({
  toolbarButton: {
    marginLeft: theme.spacing.unit
  }
});

export class Path extends React.Component {
  static propTypes = {
    classes: PropTypes.object,
    codeCombatProfile: PropTypes.any,
    match: PropTypes.object,
    onAddAssistant: PropTypes.func.isRequired,
    onAssistantKeyChange: PropTypes.func.isRequired,
    onCloseDialog: PropTypes.func.isRequired,
    onNotification: PropTypes.func.isRequired,
    onOpen: PropTypes.func.isRequired,
    onOpenSolution: PropTypes.func.isRequired,
    onProblemChangeRequest: PropTypes.func.isRequired,
    onProblemDialogShow: PropTypes.func.isRequired,
    onProblemMoveRequest: PropTypes.func.isRequired,
    onProblemSolutionSubmit: PropTypes.func.isRequired,
    onProfileUpdate: PropTypes.func.isRequired,
    onPushPath: PropTypes.func.isRequired,
    onRefreshSolutions: PropTypes.func.isRequired,
    onRemoveAssistant: PropTypes.func.isRequired,
    onRequestMoreProblems: PropTypes.func.isRequired,
    onShowCollaboratorsClick: PropTypes.func.isRequired,
    onToggleJoinStatus: PropTypes.func.isRequired,
    pathActivities: pathActivities,
    pathStatus: PropTypes.string,
    ui: PropTypes.any,
    uid: PropTypes.string
  };

  componentDidMount() {
    this.props.onOpen(this.props.match.params.pathId);
  }

  onMoveProblem = (problem, direction) => {
    const { pathActivities, onProblemMoveRequest } = this.props;

    onProblemMoveRequest(pathActivities.path.id, problem.id, direction);
  };

  onOpenProblem = problem => {
    const {
      onOpenSolution,
      onProblemSolutionSubmit,
      onPushPath,
      pathActivities
    } = this.props;
    switch (problem.type) {
      case ACTIVITY_TYPES.codeCombat.id:
      case ACTIVITY_TYPES.codeCombatNumber.id:
        onProblemSolutionSubmit(
          pathActivities.path.id,
          { problemId: problem.id, ...problem },
          "Completed"
        );
        break;
      case ACTIVITY_TYPES.jupyterInline.id:
      case ACTIVITY_TYPES.jupyter.id:
      case ACTIVITY_TYPES.youtube.id:
        onPushPath(`/paths/${pathActivities.path.id}/activities/${problem.id}`);
        break;
      default:
        onOpenSolution(pathActivities.path.id, problem);
    }
  };

  requestMoreProblems = () =>
    this.props.onRequestMoreProblems(
      this.props.uid,
      this.props.pathActivities.path.id,
      this.props.pathActivities.activities.length
    );

  refreshSolutions = () =>
    this.props.onRefreshSolutions(this.props.pathActivities.path.id);

  changeJoinStatus = () =>
    this.props.onToggleJoinStatus(
      this.props.uid,
      this.props.pathActivities.path.id,
      this.props.pathStatus === PATH_STATUS_NOT_JOINED
    );

  onAddActivityClick = () => this.props.onProblemDialogShow();
  onTextSolutionSubmit = (solution, activityId) => {
    const {
      onCloseDialog,
      onProblemSolutionSubmit,
      pathActivities
    } = this.props;
    const activity = pathActivities.activities.find(
      activity => activity.id === activityId
    );

    onProblemSolutionSubmit(
      pathActivities.path.id,
      { ...activity, problemId: activity.id },
      solution
    );
    onCloseDialog();
  };
  onProfileUpdate = profile => {
    const {
      onCloseDialog,
      onProblemSolutionSubmit,
      onProfileUpdate,
      ui
    } = this.props;

    onProfileUpdate(profile, "CodeCombat");
    onProblemSolutionSubmit(
      ui.dialog.value.path,
      { ...ui.dialog.value, problemId: ui.dialog.value.id },
      profile
    );
    onCloseDialog();
  };

  onProblemChangeRequest = (id, data) => {
    const { pathActivities = {}, onProblemChangeRequest } = this.props;
    const activities = pathActivities.activities || [];
    let additionalData = {};
    if (!data.id) {
      let maxOrderIndex = -Infinity;
      activities.forEach(activity => {
        maxOrderIndex =
          activity.orderIndex > maxOrderIndex
            ? activity.orderIndex
            : maxOrderIndex;
      });
      maxOrderIndex = maxOrderIndex === -Infinity ? 1 : maxOrderIndex + 1;
      additionalData = {
        orderIndex: maxOrderIndex
      };
    }
    onProblemChangeRequest(id, {
      ...data,
      ...additionalData
    });
  };
  render() {
    const {
      classes,
      codeCombatProfile,
      match,
      onAddAssistant,
      onAssistantKeyChange,
      onCloseDialog,
      onProblemDialogShow,
      onShowCollaboratorsClick,
      onRemoveAssistant,
      pathActivities,
      pathStatus,
      ui,
      uid
    } = this.props;

    if (!uid) {
      return <div>Login required to display this page</div>;
    }

    if (!(pathActivities && pathActivities.path)) {
      return <LinearProgress />;
    }

    const allFinished =
      (pathActivities.activities || []).filter(problem => problem.solved)
        .length === (pathActivities.activities || []).length;
    let pathName;

    if (match.params.pathId[0] !== "-") {
      pathName = "Default";
    }

    pathName =
      pathName || (pathActivities.path && pathActivities.path.name) || "";

    return (
      <Fragment>
        <Breadcrumbs
          action={
            pathStatus !== PATH_STATUS_OWNER && [
              allFinished && {
                label: "Request more",
                handler: this.requestMoreProblems.bind(this)
              },
              !allFinished && {
                label: "Refresh",
                handler: this.refreshSolutions.bind(this)
              },
              {
                label: pathStatus === PATH_STATUS_JOINED ? "Leave" : "Join",
                handler: this.changeJoinStatus.bind(this)
              }
            ]
          }
          paths={[
            {
              label: "Paths",
              link: "/paths"
            },
            {
              label: pathName
            }
          ]}
        />
        {pathStatus === PATH_STATUS_OWNER &&
          (!APP_SETTING.isSuggesting ? (
            <Toolbar>
              <Button
                color="primary"
                onClick={this.onAddActivityClick}
                variant="raised"
              >
                Add Activity
              </Button>
              <Button
                className={classes.toolbarButton}
                onClick={() => onShowCollaboratorsClick(pathActivities.path.id)}
                variant="raised"
              >
                Collaborators
              </Button>
            </Toolbar>
          ) : (
            <Button
              aria-label="Add"
              color="primary"
              onClick={this.onAddActivityClick}
              style={{
                position: "fixed",
                bottom: 20,
                right: 20
              }}
              variant="fab"
            >
              <AddIcon />
            </Button>
          ))}
        <AddTextSolutionDialog
          onClose={onCloseDialog}
          onCommit={this.onTextSolutionSubmit}
          open={ui.dialog.type === `${ACTIVITY_TYPES.text.id}Solution`}
          solution={ui.dialog.solution}
          taskId={ui.dialog.value && ui.dialog.value.id}
        />
        <AddJestSolutionDialog
          onClose={onCloseDialog}
          onCommit={this.onTextSolutionSubmit}
          open={ui.dialog.type === `${ACTIVITY_TYPES.jest.id}Solution`}
          problem={ui.dialog.value}
          taskId={ui.dialog.value && ui.dialog.value.id}
        />
        <AddProfileDialog
          defaultValue={(codeCombatProfile && codeCombatProfile.id) || ""}
          externalProfile={{
            url: "https://codecombat.com",
            id: "CodeCombat",
            name: "Code Combat",
            description: "learn to Code JavaScript by Playing a Game"
          }}
          onClose={onCloseDialog}
          onCommit={this.onProfileUpdate}
          open={ui.dialog.type === `${ACTIVITY_TYPES.profile.id}Solution`}
        />
        <ActivitiesTable
          activities={pathActivities.activities || []}
          currentUserId={uid || "Anonymous"}
          onEditProblem={onProblemDialogShow}
          onMoveProblem={this.onMoveProblem}
          onOpenProblem={this.onOpenProblem}
          pathOwnerId={pathActivities.path && pathActivities.path.owner}
          selectedPathId={(pathActivities.path && pathActivities.path.id) || ""}
        />
        <ActivityDialog
          activity={ui.dialog.value}
          onClose={onCloseDialog}
          onCommit={this.onProblemChangeRequest}
          open={ui.dialog.type === "ProblemChange"}
          pathId={(pathActivities.path && pathActivities.path.id) || ""}
          uid={uid || "Anonymous"}
        />
        <ControlAssistantsDialog
          assistants={ui.dialog && ui.dialog.assistants}
          newAssistant={ui.dialog && ui.dialog.newAssistant}
          onAddAssistant={onAddAssistant}
          onAssistantKeyChange={onAssistantKeyChange}
          onClose={onCloseDialog}
          onRemoveAssistant={onRemoveAssistant}
          open={ui.dialog.type === "CollaboratorsControl"}
          target={pathActivities.path && pathActivities.path.id}
        />
      </Fragment>
    );
  }
}

sagaInjector.inject(sagas);

const mapStateToProps = (state, ownProps) => ({
  pathActivities: pathActivitiesSelector(state, ownProps),
  pathStatus: pathStatusSelector(state, ownProps),
  ui: state.path.ui,
  uid: state.firebase.auth.uid,
  codeCombatProfile: codeCombatProfileSelector(state)
});

const mapDispatchToProps = {
  onAddAssistant: pathAddCollaboratorRequest,
  onAssistantKeyChange: assignmentAssistantKeyChange,
  onCloseDialog: pathCloseDialog,
  onNotification: notificationShow,
  onOpen: pathOpen,
  onShowCollaboratorsClick: pathShowCollaboratorsDialog,
  onProfileUpdate: externalProfileUpdateRequest,
  onOpenSolution: pathOpenSolutionDialog,
  onProblemChangeRequest: pathProblemChangeRequest,
  onProblemDialogShow: pathProblemDialogShow,
  onProblemMoveRequest: pathActivityMoveRequest,
  onProblemSolutionSubmit: problemSolutionSubmitRequest,
  onPushPath: push,
  onRefreshSolutions: pathRefreshSolutionsRequest,
  onRemoveAssistant: pathRemoveCollaboratorRequest,
  onRequestMoreProblems: pathMoreProblemsRequest,
  onToggleJoinStatus: pathToggleJoinStatusRequest
};

export default compose(
  withStyles(styles),
  withRouter,
  firebaseConnect((ownProps, store) => {
    const state = store.getState();
    const uid = state.firebase.auth.uid;
    const pathId = ownProps.match.params.pathId;

    if (!uid) {
      return false;
    }

    return [
      `/completedActivities/${uid}/${pathId}`,
      `/paths/${pathId}`,
      {
        path: "/activities",
        queryParams: ["orderByChild=path", `equalTo=${pathId}`]
      },
      `/studentJoinedPaths/${uid}/${pathId}`,
      `userAchievements/${uid}/CodeCombat`
    ];
  }),
  connect(mapStateToProps, mapDispatchToProps)
)(Path);
