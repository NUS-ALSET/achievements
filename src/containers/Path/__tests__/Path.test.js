import React from "react";
import { createShallow } from "@material-ui/core/test-utils";
import { Path } from "../Path";

// noinspection RequiredAttributes
describe("<Path>", () => {
  let shallow;
  let spy;

  beforeEach(() => {
    shallow = createShallow();
    spy = jest.fn();
  });

  it("should check snapshot", () => {
    const wrapper = shallow(
      <Path
        match={{
          params: {
            pathId: "testPath"
          }
        }}
        onAddAssistant={spy}
        onAssistantKeyChange={spy}
        onCloseDialog={spy}
        onNotification={spy}
        onOpen={spy}
        onOpenSolution={spy}
        onProblemChangeRequest={spy}
        onProblemDialogShow={spy}
        onProblemMoveRequest={spy}
        onProblemSolutionSubmit={spy}
        onProfileUpdate={spy}
        onPushPath={spy}
        onRemoveAssistant={spy}
        onRequestMoreProblems={spy}
        onShowCollaboratorsClick={spy}
        onToggleJoinStatus={spy}
        pathActivities={{
          activities : []
        }}
        pathStatus="test"
        ui={{
          dialog: false
        }}
        uid="testUser"
      />
    );

    expect(wrapper).toMatchSnapshot();
  });
});
