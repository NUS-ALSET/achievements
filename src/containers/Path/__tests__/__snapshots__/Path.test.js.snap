// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Path> should check snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Path
    match={
        Object {
            "params": Object {
              "pathId": "testPath",
            },
          }
    }
    onAddAssistant={[Function]}
    onAssistantKeyChange={[Function]}
    onCloseDialog={[Function]}
    onNotification={[Function]}
    onOpen={[Function]}
    onOpenSolution={[Function]}
    onProblemChangeRequest={[Function]}
    onProblemDialogShow={[Function]}
    onProblemMoveRequest={[Function]}
    onProblemSolutionSubmit={[Function]}
    onProfileUpdate={[Function]}
    onPushPath={[Function]}
    onRemoveAssistant={[Function]}
    onRequestMoreProblems={[Function]}
    onShowCollaboratorsClick={[Function]}
    onToggleJoinStatus={[Function]}
    pathActivities={
        Object {
            "activities": Array [],
          }
    }
    pathStatus="test"
    ui={
        Object {
            "dialog": false,
          }
    }
    uid="testUser"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {},
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {},
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "context": Object {},
  },
}
`;
